---
title: Welcome to Apex
author: 
- name: Nick Borcherding
  email: ncborch@gmail.com
  affiliation: Washington University in St. Louis, School of Medicine, St. Louis, MO, USA
date: "April 26, 2024"
output:
  BiocStyle::html_document:
    toc_float: true
package: Trex
vignette: >
  %\VignetteEngine{knitr::knitr}
  %\VignetteIndexEntry{Running Apex}
  %\usepackage[UTF-8]{inputenc}
---

```{r, echo=FALSE, results="hide", message=FALSE}
knitr::opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE)
library(BiocStyle)
```

# Introduction

## Load Libraries

```{r}
suppressMessages(library(Apex))
suppressMessages(library(keras))
```

Apex is meant to serve as an API for deep learning models based on immune receptor sequencing. These collection of functions serve to process amino acid sequences, 

****************

# Generate.sequences

```{r}
sequences <- generate.sequences(prefix.motif = "CAS",
                                suffix.motif = "YF",
                                number.of.sequences = 1000,
                                min.length = 8,
                                max.length = 16)
head(sequences)
```


# Mutate.sequences

```{r}
mutated_sequences <- mutate.sequences(sequences, 
                                      n.sequence = 1,
                                      position.start = 3,                                  
                                      position.end = 8)
head(sequences)
head(mutated_sequences)
```
# Encoders

## One.hot.encoder

#' @param input.sequences The amino acid sequences to use
#' @param max.length Additional length to pad, NULL will pad sequences 
#' to the max length of input.sequences
#' @param convert.to.matrix Return a matrix (TRUE) or a 3D array (FALSE)

```{r}
sequence.matrix <- one.hot.encoder(input.sequences =  c(sequences, mutated_sequences), 
                                   convert.to.matrix = TRUE)
```

## Property.encoder

Encode **input.sequences** into an array/matrix based on amino acid properties. 

**method.to.use**  

* atchleyFactors - \href{https://pubmed.ncbi.nlm.nih.gov/15851683/}{citation}).
* crucianiProperties - \href{https://analyticalsciencejournals.onlinelibrary.wiley.com/doi/abs/10.1002/cem.856}{citation}).
* FASGAI - \href{https://pubmed.ncbi.nlm.nih.gov/18318694/}).
* kideraFactors - \href{https://link.springer.com/article/10.1007/BF01025492}{citation}).
* MSWHIM - \href{https://pubs.acs.org/doi/10.1021/ci980211b}{citation}).
* ProtFP - \href{https://pubmed.ncbi.nlm.nih.gov/24059694/}{citation}).
* stScales - \href{https://pubmed.ncbi.nlm.nih.gov/19373543/}{citation}).
* tScales - \href{https://www.sciencedirect.com/science/article/abs/pii/S0022286006006314}{citation}).
* VHSE - \href{https://pubmed.ncbi.nlm.nih.gov/15895431/}{citation}).
* zScales - \href{https://pubmed.ncbi.nlm.nih.gov/9651153/}{citation}).


```{r}
property.matrix <- property.encoder(input.sequences =  c(sequences, mutated_sequences), 
                                    method.to.use = "FASGAI",
                                    convert.to.matrix = TRUE)

head(property.matrix[,1:20])
```

## Positional.encoder

```{r}
positional.matrix <- positional.encoder(number.of.sequences = length(c(sequences, mutated_sequences)),
                                        latent.dims = 40)
```

## 
***
# Conclusion

This has been a general overview of the capabilities of Trex for incorporating TCR information into the embedding space of single-cell data. If you have any questions, comments, or suggestions, feel free to visit the github repository or [email me](mailto:ncborch@gmail.com).

## Session Info

```{r}
sessionInfo()
```
